test-framework/sudo-compliance-tests/src/pam/env.rs

```diff
@@ -0,0 +1,303 @@
+    .file(ETC_ENVIRONMENT_PATH, format!("{set_name}={set_value}"))
+    .file(
+        SECURITY_PAM_ENV_PATH,
+        [
+            STOCK_SECURITY_PAM_ENV,
+            &format!("{default_name} DEFAULT={default_value}"),
+            &format!("{override_name} OVERRIDE={override_value}"),
+        ],
+    )
+    .build()?;
+
+    let new_set_value = "0";
+    let new_default_value = "1";
+    let new_override_value = "2";
+    let stdout = Command::new("env")
+        .arg(format!("{set_name}={new_set_value}"))
+        .arg(format!("{default_name}={new_default_value}"))
+        .arg(format!("{override_name}={new_override_value}"))
+        .args(["sudo", "env"])
+        .exec(&env)?
+        .stdout()?;
+    let env = helpers::parse_env_output(&stdout)?;
+
+    assert_eq!(Some(new_set_value), env.get(set_name).copied());
+    assert_eq!(Some(new_default_value), env.get(default_name).copied());
+    assert_eq!(Some(new_override_value), env.get(override_name).copied());
+
+    Ok(())
+}
+
+#[test]
+fn env_keep_has_precedence_over_pam_env() -> Result<()> {
+    env_list_has_precendece_over_pam_env(EnvList::Keep)
+}
+
+#[test]
+fn env_check_has_precedence_over_pam_env() -> Result<()> {
+    env_list_has_precendece_over_pam_env(EnvList::Check)
+}
+
+#[test]
+#[ignore = "gh420"]
+fn var_rejected_by_env_check_falls_back_to_pam_env_value() -> Result<()> {
+    let set_name = "SET_VAR";
+    let set_value = "set";
+    let default_name = "DEFAULT_VAR";
+    let default_value = "default";
+    let override_name = "OVERRIDE_VAR";
+    let override_value = "override";
+    let env = Env([
+        SUDOERS_ALL_ALL_NOPASSWD,
+        &format!("Defaults env_check = \"{set_name} {default_name} {override_name}\""),
+    ])
+    .file(PAM_D_SUDO_PATH, [STOCK_PAM_D_SUDO, PAM_D_SUDO_READENV])
+    .file(ETC_ENVIRONMENT_PATH, format!("{set_name}={set_value}"))
+    .file(
+        SECURITY_PAM_ENV_PATH,
+        [
+            STOCK_SECURITY_PAM_ENV,
+            &format!("{default_name} DEFAULT={default_value}"),
+            &format!("{override_name} OVERRIDE={override_value}"),
+        ],
+    )
+    .build()?;
+
+    let new_set_value = "%0";
+    let new_default_value = "%1";
+    let new_override_value = "%2";
+    let stdout = Command::new("env")
+        .arg(format!("{set_name}={new_set_value}"))
+        .arg(format!("{default_name}={new_default_value}"))
+        .arg(format!("{override_name}={new_override_value}"))
+        .args(["sudo", "env"])
+        .exec(&env)?
+        .stdout()?;
+    let env = helpers::parse_env_output(&stdout)?;
+
+    assert_eq!(Some(set_value), env.get(set_name).copied());
+    assert_eq!(Some(default_value), env.get(default_name).copied());
+    assert_eq!(Some(override_value), env.get(override_name).copied());
+
+    Ok(())
+}
+
+#[test]
+#[ignore = "gh420"]
+fn default_and_override_pam_env_vars_are_parentheses_checked_but_set_vars_are_not() -> Result<()> {
```

Comment:
This is done regardless of `env_check` being set on the sudoers file?

Reply:
there's another list of env vars that should be preserved: `env_keep`. `env_keep` does not check for `%` and `/` characters like `env_check` does.

however, the variables in either list are subjected to this check: variables whose values start with `()` are not preserved. this parentheses check appears to apply to some variables set by `pam_env.so` depending on what kind of variable they are (here "kind" refers to where they are defined: `/etc/environment` vs `/etc/security/pam_env.conf`).